src/libsyntax/ext/tt/macro_check.rs

```diff
@@ -0,0 +1,342 @@
+use crate::symbol::{kw, sym};
+
+use rustc_data_structures::fx::FxHashMap;
+use syntax_pos::{symbol::Ident, Span};
+
+/// Information attached to a meta-variable binder in LHS.
+struct BinderInfo {
+    /// The span of the meta-variable in LHS.
+    span: Span,
+    /// The stack of Kleene operators (outermost first).
+    ops: Vec<KleeneToken>,
+}
+
+/// An environment of meta-variables to their binder information.
+type Binders = FxHashMap<Ident, BinderInfo>;
+
+/// The state at which we entered a macro definition in the RHS of another macro definition.
+#[derive(Clone)]
+struct MacroState<'a> {
+    /// The binders of the branch where we entered the macro definition.
+    binders: &'a Binders,
+    /// The stack of Kleene operators (outermost first) where we entered the macro definition.
+    ops: &'a [KleeneToken],
+}
+
+/// Checks that meta-variables are bound and used correctly for each left and right hand-side pairs
+/// of a macro definition. The span is used when no span information is available. In presence of
+/// nested macro definitions (a macro that defines a macro), this check may have false positives and
+/// false negatives.
+pub fn check_meta_variables(
+    sess: &ParseSess,
+    span: Span,
+    lhses: &[TokenTree],
+    rhses: &[TokenTree],
+) -> bool {
+    if lhses.len() != rhses.len() {
+        sess.span_diagnostic.span_bug(span, "length mismatch between LHSes and RHSes")
+    }
+    let mut valid = true;
+    for (i, lhs) in lhses.iter().enumerate() {
+        let rhs = &rhses[i];
+        let mut binders = Binders::default();
+        check_binders(sess, lhs, &[], &mut binders, &[], &mut valid);
+        check_occurrences(sess, rhs, &[], &binders, &[], &mut valid);
+    }
+    valid
+}
+
+/// Checks that `lhs` does not contain duplicate binders (e.g. `$a:expr, $a:expr` since it would be
+/// ambiguous which `$a` to use if we ever needed to) according to `binders` and `macros`. Extends
+/// `binders` with new binders according to `ops`, the stack of Kleene operators. Sets `valid` to
+/// false if an error is encountered.
+fn check_binders(
+    sess: &ParseSess,
+    lhs: &TokenTree,
+    macros: &[MacroState<'_>],
+    binders: &mut Binders,
+    ops: &[KleeneToken],
+    valid: &mut bool,
+) {
+    match *lhs {
+        TokenTree::Token(..) => {}
+        TokenTree::MetaVar(span, name) => {
+            if macros.is_empty() {
+                sess.span_diagnostic.span_bug(span, "unexpected MetaVarDecl in rhs");
+            }
+            if get_binder_info(macros, binders, name).is_none() {
+                binders.insert(name, BinderInfo { span, ops: ops.to_owned() });
+            } else {
+                check_occurrences(sess, lhs, macros, binders, ops, valid);
+            }
+        }
+        TokenTree::MetaVarDecl(span, name, _kind) => {
+            if let Some(prev_info) = get_binder_info(macros, binders, name) {
+                // TODO: If macros.len() > 0, notify that it may be a false positive.
```

Comment:
Actually, I think this lint will also have false positives in cases we can't detect. For example:
```
macro_rules! foo {
    ($name:ident) => { $name! bar { ($x:ident) => { $x }; } };
}
foo!(macro_rules);
```
So I'm removing those TODOs since they could be misleading by thinking that when they are not present, the lint is complete (I'll update the description of the PR). When migrating to lint, if I find a way to add this information to the lint I will.